### ISSUE TITLE
Faster R-CNN for Text Detection

### ISSUE THREAD

**Issue by alexattia** [yhenon/keras-frcnn#82]
I have trained the network for text detection (only for digits) and recognition with multiple thousands of pictures. My pictures are like this one (I am only predicting the clock, 21:08 on this picture) but in addition, I have different kinds of scoreboards:  
![test_4](https://user-images.githubusercontent.com/12381760/27572577-a1437b72-5ac1-11e7-8ec9-76f142887745.png)
After a couple of hours of training, it doesn't work at all on a test set. Do you have any idea what I can do?

**Comment by alexiskattan** [comment]
sounds interesting would like to hear what other people thing.  Also where did you get the labeled data?

**Comment by shahruk10** [comment]
If your regions of interests are small (like < 200 px) you can use anchor boxes of smaller size (like 50 and 150) and try out ratios that fit your bounding boxes ( 2:1 ~ 3:1) . These are found in the config.py file.

**Comment by alexiskattan** [comment]
I'm having similar issues getting printing `all_dets` after training it on a small sample data set.

**Comment by yhenon** [comment]
Please share your training log, it's hard to know anything without it

**Comment by alexattia** [comment]
@shahruk10 Thanks, I will try with different ratios, probably 3:1.
@yhenon During training, this exception occurred multiple times: `Exception: a must be non-empty`. I don't know if it's important.   
At the end of training (~ 6 hours), the losses are like this:  
```
Loss RPN classifier: 0.002610719088972545
Loss RPN regression: 0.082112987785389122
Loss Detector classifier: 0.2533039738115739
Loss Detector regression: 0.12925621128920464
Elapsed time: 2528.9011764526367
Total loss decreased from 0.497426369193454 to 0.4672838919751402, saving weights
```
Thank you very much

**Comment by yhenon** [comment]
These look ok. further questions:
- how many tagged images do you have?
- how many classes?

**Comment by alexattia** [comment]
I have 11 classes including bg and the tagged images distribution is : 
```
{'0': 975,
 '1': 1275,
 '2': 1226,
 '3': 1289,
 '4': 1187,
 '5': 1285,
 '6': 795,
 '7': 771,
 '8': 684,
 '9': 510,
 'bg': 0}
```

**Comment by yhenon** [comment]
It's somewhat surprising you see no detections then - losses look ok, and dataset seems adequate too. What do you mean by 'doesnt work at all`? Is it that you see no predictions, or that the predictions are wrong?

**Comment by alexattia** [comment]
I see no predictions : I tried on 15 pictures (different times, different scoreboards types) and there isn't any prediction for every picture.   
I am currently training with 3/1 ratios and I added 64px to anchor scales.

**Comment by yhenon** [comment]
Given how low the RPN losses are, I wouldn't expect too much from that. Could you try lowering the `bbox_threshold` to a lower value (default is 0.8, maybe try 0.2 or 0.5) in test_frcnn.py and see if that works?

**Comment by alexattia** [comment]
For `bbox_threshold == 0.5`, it doesn't change anything : no prediction.  
For  `bbox_threshold == 0.2`, I have some predictions ! It was a good idea. But the probabilities are very low : 
`[('2', 24.095310270786285), ('1', 22.178661823272705)]`
![test_2](https://user-images.githubusercontent.com/12381760/27603824-3f34d226-5b2c-11e7-9b9c-dc9ff885e830.png)
What do you think I can do to improve this ?

**Comment by yhenon** [comment]
It seems like the classifier cant distinguish in between classes - it looks like you're tagging individual digits? 
I don't know if faster-rcnn is quite the right tool for this job since (a) the resolution might somewhat low (there is a pooling of 16x before the input to the classifier layers) and (b) boxes from the RPN might overlap the digits poorly. There is also the issue of digits present that are not part of the scoreboard, that need to be learned as being bg regions.

**Comment by alexattia** [comment]
Yes, I am tagging individual digits.  
You are probably right, not sure that faster rcnn is the right tool for this. I tried with HD pictures (less data) and it only recognized ':' (as a digit ..) in the clock.

**Comment by JieZou1** [comment]
@yhenon 

I am also working on detecting texts on images, and thinking about using R-CNN based methods. It is very interesting reading your comments on 6/27. 

Why do you think the classifier can't distinguish between classes? For this kind of digits classification, the deep NN classifier could do it very well, right? Then, why couldn't RCNN classifier handle it well? I understand that 16x pooling may certainly be too big, but there are other pooling methods, right?

What do you mean boxes from the RPN might overlap the digits poorly? If we could define the anchor sizes and aspect ratios right to the digits in the image, it is the same regression problem as usual (big) object detection, right?

In worse case, can we pool all digits together as one foreground class? In this case, we basically need RPN only. It is still very useful for detecting candidates. These candidates can then be classified by a standalone NN classifier. For this kind of simple digits, the NN classifier should achieve at least 99% accuracy. And, it does not required to be a very deep NN (2-3 Conv Layers should be more than enough), so the speed may not be a critical issue. What do you think this approach (use RPN for candidate detection and then use a conv net for classification)?

Many thanks indeed for this wonderful implementation of Faster R-CNN. It certainly helps a lot for me to understand the algorithm.
